src/boot/efifirmware.c

```diff
@@ -0,0 +1,60 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include "efifirmware.h"
+#include "util.h"
+#include <endian.h>
+
+static bool efifw_validate_header(const void *blob, size_t uki_efifw_len,
+                                  uint32_t *fwid_off, uint32_t *payload_off) {
+        if ((uintptr_t) blob % alignof(EfiFwHeader) != 0)
+                return false;
+        const EfiFwHeader *fw_header = ASSERT_PTR(blob);
+
+        if (le32toh(fw_header->base.magic) != FWHEADERMAGIC)
+                return false;
+
+        size_t payload_len = le32toh(fw_header->base.payload_len);
+        size_t header_len = le32toh(fw_header->base.header_len);
+        size_t fwid_len = le32toh(fw_header->base.fwid_len);
+
+         /* at least entire base header must be present */
+        if (header_len < sizeof (fw_header->base))
+                return false;
+
+        size_t fw_base_size = header_len + fwid_len + payload_len;
+
+        if (uki_efifw_len < sizeof(EfiFwHeader) ||
+            uki_efifw_len < fw_base_size)
+                return false;
+
+        if (PTR_TO_SIZE(blob) > SIZE_MAX - fw_base_size)
+                return false;
+
+        if (!ADD_SAFE(fwid_off, PTR_TO_SIZE(blob), header_len))
+                return false;
+
+        if (!ADD_SAFE(payload_off, *fwid_off, fwid_len))
+                return false;
```

Comment:
we have this rule in system that return params should be initialized unconditionally on success, and not be touched whatsoever on failure. in this function you might fail but still initialize return params (i.e. this 2nd AFF_SAFE fails, then you already initialized fwid_off...). please follow style on this, i.e. do all calculations first, and only once everything succeeded copy to return params and return success.